rabbitmq/rabbitmq-server
# Node drain (maintenance mode) operations

## Proposed Changes

This introduce a concept of maintenance mode together with initial implementations of `drain` and `revive` commands.
See #2321 for the background and details.

This ability is hidden behind a new feature flag, `maintenance_mode_status`. While this node
is mostly node-local, it depends on a new table and if we add it on node boot unconditionally, it will
fail schema consistency checks in mixed-version clusters.

Depends on

 * https://github.com/rabbitmq/rabbitmq-common/pull/399
 * https://github.com/rabbitmq/rabbitmq-cli/pull/419
 * https://github.com/rabbitmq/rabbitmq-web-dispatch/pull/43
 * https://github.com/rabbitmq/rabbitmq-web-stomp/pull/125
 * https://github.com/rabbitmq/rabbitmq-stomp/pull/149
 * https://github.com/rabbitmq/rabbitmq-mqtt/pull/234
 * https://github.com/rabbitmq/rabbitmq-web-mqtt/pull/67

## How to QA with All Compatible Nodes

Start a three-node cluster where one node has multiple protocol plugins (some examples are provided below).

Then declare some classic and quorum queues, make the classic queues mirrored and open a few connections to a node from a REPL. Then put nodes into maintenance mode using `rabbit_maintenance:drain/0` and `rabbit_maintenance:revive/0` from their shells. Observe what is logged, how queue master replicas migrate, try opening new connections to a node in maintenance mode, and so on. #2321 explains what the `drain` command currently does.

``` shell
%% see peer_discovery.classic_config.1.conf
gmake run-broker PLUGINS="rabbitmq_management rabbitmq_amqp1_0 rabbitmq_mqtt rabbitmq_stomp rabbitmq_web_mqtt rabbitmq_web_stomp" RABBITMQ_CONFIG_FILE=/path/to/peer_discovery.classic_config.1.conf

%% see peer_discovery.classic_config.2.conf
gmake run-broker PLUGINS="rabbitmq_management" RABBITMQ_NODENAME=hare@localhost RABBITMQ_NODE_PORT=5673 RABBITMQ_DIST_PORT=25673 RABBITMQ_CONFIG_FILE=/path/to/peer_discovery.classic_config.2.conf

%% see peer_discovery.classic_config.3.conf
gmake run-broker PLUGINS="rabbitmq_management" RABBITMQ_NODENAME=flopsy@localhost RABBITMQ_NODE_PORT=5674 RABBITMQ_DIST_PORT=25674 RABBITMQ_CONFIG_FILE=/path/to/peer_discovery.classic_config.3.conf
```

``` ini
# peer_discovery.classic_config.1.conf
cluster_formation.peer_discovery_backend = rabbit_peer_discovery_classic_config

cluster_formation.classic_config.nodes.1 = rabbit@localhost
cluster_formation.classic_config.nodes.2 = hare@localhost
cluster_formation.classic_config.nodes.3 = flopsy@localhost

cluster_formation.randomized_startup_delay_range.min = 2
cluster_formation.randomized_startup_delay_range.max = 10

log.file.level = debug

listeners.tcp.1 = 127.0.0.1:5672
listeners.tcp.2 = :::5672
```

``` ini
# peer_discovery.classic_config.2.conf
cluster_formation.peer_discovery_backend = rabbit_peer_discovery_classic_config

cluster_formation.classic_config.nodes.1 = rabbit@localhost
cluster_formation.classic_config.nodes.2 = hare@localhost
cluster_formation.classic_config.nodes.3 = flopsy@localhost

cluster_formation.randomized_startup_delay_range.min = 3
cluster_formation.randomized_startup_delay_range.max = 6

log.file.level = debug

listeners.tcp.1 = 127.0.0.1:5673
listeners.tcp.2 = :::5673

management.tcp.port = 15673
```

``` ini
# peer_discovery.classic_config.3.conf
cluster_formation.peer_discovery_backend = rabbit_peer_discovery_classic_config

cluster_formation.classic_config.nodes.1 = rabbit@localhost
cluster_formation.classic_config.nodes.2 = hare@localhost
cluster_formation.classic_config.nodes.3 = flopsy@localhost

cluster_formation.randomized_startup_delay_range.min = 2
cluster_formation.randomized_startup_delay_range.max = 10

log.file.level = debug

listeners.tcp.1 = 127.0.0.1:5674
listeners.tcp.2 = :::5674

management.tcp.port = 15684
```

## How to QA with In a Mixed Version Cluster

 * Create a cluster of two `3.8.5` nodes (sufficient for this test), A and B
 * Declare a basic topology
 * Stop both nodes
 * Copy their data directories
 * Start node A
 * Start a node from source with data directory from node B (node B's data dir can be copied to `$TMPDIR/rabbitmq-test-instances/rabbit@{hostname}/mnesia/rabbit@{hostname}`, for example)
 * Inspect feature flags
 * Run `ERL_LIBS=../rabbit/plugins ./escript/rabbitmq-diagnostics cluster_status`
 * Stop node A
 * Start another node from source with node A's data directory
 * Inspect feature flags
 * Try draining and reviving either node


## Types of Changes

- [ ] Bug fix (non-breaking change which fixes issue #2321)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)

## Checklist

- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [x] Any dependent changes have been merged and published in related repositories
